pkg/promtail/targets/pubsub/targetmanager.go

```diff
@@ -0,0 +1,184 @@
+package pubsub
+
+import (
+	"context"
+	"fmt"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/grafana/loki/pkg/promtail/api"
+	"github.com/grafana/loki/pkg/promtail/scrapeconfig"
+	"github.com/grafana/loki/pkg/promtail/targets/target"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/relabel"
+	"google.golang.org/api/option"
+)
+
+var (
+// TODO(kavi): metrics for pubsub target
+)
+
+type PubsubTargetManager struct {
+	logger  log.Logger
+	targets map[string]*PubsubTarget
+}
+
+func NewPubsubTargetManager(
+	logger log.Logger,
+	client api.EntryHandler,
+	scrape []scrapeconfig.Config,
+) (*PubsubTargetManager, error) {
+	tm := &PubsubTargetManager{
+		logger:  logger,
+		targets: make(map[string]*PubsubTarget),
+	}
+
+	for _, cf := range scrape {
+		t, err := NewPubsubTarget(logger, client, cf.RelabelConfigs, cf.JobName, cf.PubsubConfig)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create pubsub target: %w", err)
+		}
+		tm.targets[cf.JobName] = t
+	}
+
+	return tm, nil
+}
+
+func (tm *PubsubTargetManager) Ready() bool {
+	for _, t := range tm.targets {
+		if t.Ready() {
+			return true
+		}
+	}
+	return false
+}
+
+func (tm *PubsubTargetManager) Stop() {
+	for name, t := range tm.targets {
+		if err := t.Stop(); err != nil {
+			level.Error(t.logger).Log("event", "failed to stop pubsub target", "name", name, "cause", err)
+		}
+	}
+}
+
+func (tm *PubsubTargetManager) ActiveTargets() map[string][]target.Target {
+	// TODO(kavi): if someway to check if specific topic is active and store the state on the target struct?
+	return tm.AllTargets()
+}
+
+func (tm *PubsubTargetManager) AllTargets() map[string][]target.Target {
+	res := make(map[string][]target.Target, len(tm.targets))
+	for k, v := range tm.targets {
+		res[k] = []target.Target{v}
+	}
+	return res
+}
+
+type PubsubTarget struct {
+	logger        log.Logger
+	handler       api.EntryHandler
+	config        *scrapeconfig.PubsubTargetConfig
+	relabelConfig []*relabel.Config
+	jobName       string
+
+	// lifecycle management
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	// pubsub
+	ps   *pubsub.Client
+	msgs chan *pubsub.Message
+}
+
+func NewPubsubTarget(
+	logger log.Logger,
+	handler api.EntryHandler,
+	relabel []*relabel.Config,
+	jobName string,
+	config *scrapeconfig.PubsubTargetConfig,
+) (*PubsubTarget, error) {
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	ps, err := pubsub.NewClient(ctx, config.ProjectID, option.WithCredentialsFile(config.CredentialsPath))
+	if err != nil {
+		return nil, err
+	}
+
+	pt := &PubsubTarget{
+		logger:        logger,
+		handler:       handler,
```

Comment:
I think the config contains pipeline stages that you need to build and hook with the client.

Reply:
Ah. I see. nice catch! will do 👍